src/pharmpy/modeling/eta_additions.py

```diff
@@ -203,12 +332,45 @@ def add_iov(model, occ, list_of_parameters=None, eta_names=None):
 
         sset.subs({eta.name: S(f'ETAI{i}')})
 
-    iovs.extend(etais)
-    iovs.extend(sset)
+    return iovs, etais
 
-    model.random_variables, model.parameters, model.statements = rvs, pset, iovs
 
-    return model
+def _add_iov_etas_disjoint(rvs, pset, etas, indices, categories, omega_iov_name, eta_name):
+
+    _add_iov_declare_diagonal_omegas(rvs, pset, etas, indices, omega_iov_name)
+
+    for i in indices:
+        omega_iov = S(omega_iov_name(i, i))
+        for k in range(1, len(categories) + 1):
+            yield RandomVariable.normal(eta_name(i, k), 'iov', 0, omega_iov)
+
+
+def _add_iov_etas_joint(rvs, pset, etas, indices, categories, omega_iov_name, eta_name):
+
+    _add_iov_declare_diagonal_omegas(rvs, pset, etas, indices, omega_iov_name)
+
+    # NOTE Declare off-diagonal OMEGAs
+    for i, j in combinations(indices, r=2):
+        omega_iov = S(omega_iov_name(i, j))
+        omega_iiv = rvs.get_covariance(etas[i - 1], etas[j - 1])
+        init = pset[omega_iiv].init * 0.1 if omega_iiv != 0 and omega_iiv in pset else 0.001
+        pset.append(Parameter(str(omega_iov), init=init))
```

Comment:
If `get_covariance(i,i)` was equivalent to `get_variance(i)` we could have a single implementation for diagonal and off-diagonal.